File: src/test/java/org/snakeyaml/engine/usecases/env/EnvVariableTest.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.snakeyaml.engine.usecases.env;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.snakeyaml.engine.v2.api.Load;
import org.snakeyaml.engine.v2.api.LoadSettings;
import org.snakeyaml.engine.v2.env.EnvConfig;
import org.snakeyaml.engine.v2.exceptions.MissingEnvironmentVariableException;
import org.snakeyaml.engine.v2.utils.TestUtils;

@org.junit.jupiter.api.Tag("fast")
class EnvVariableTest {

  // the variables EnvironmentKey1 and EnvironmentEmpty are set by Maven
  private static final String KEY1 = "EnvironmentKey1";
  private static final String EMPTY = "EnvironmentEmpty";
  private static final String VALUE1 = "EnvironmentValue1";

  @Test
  @DisplayName("Parse docker-compose.yaml example")
  public void testDockerCompose() {
    Load loader =
        new Load(LoadSettings.builder().setEnvConfig(Optional.of(new EnvConfig() {})).build());
    String resource = TestUtils.getResource("env/docker-compose.yaml");
    Map<String, Object> compose = (Map<String, Object>) loader.loadFromString(resource);
    String output = compose.toString();
    assertTrue(output.endsWith(
        "environment={URL1=EnvironmentValue1, URL2=, URL3=server3, URL4=, URL5=server5, URL6=server6}}}}"),
        output);
  }

  @Test
  @DisplayName("Custom EVN config example")
  public void testCustomEnvConfig() {
    HashMap<String, String> provided = new HashMap();
    provided.put(KEY1, "VVVAAA111");
    System.setProperty(EMPTY, "VVVAAA222");
    Load loader = new Load(
        LoadSettings.builder().setEnvConfig(Optional.of(new CustomEnvConfig(provided))).build());
    String resource = TestUtils.getResource("env/docker-compose.yaml");
    Map<String, Object> compose = (Map<String, Object>) loader.loadFromString(resource);
    String output = compose.toString();
    assertTrue(output.endsWith(
        "environment={URL1=VVVAAA111, URL2=VVVAAA222, URL3=VVVAAA222, URL4=VVVAAA222, URL5=server5, URL6=server6}}}}"),
        output);
  }

  private String load(String template) {
    Load loader =
        new Load(LoadSettings.builder().setEnvConfig(Optional.of(new EnvConfig() {})).build());
    String loaded = (String) loader.loadFromString(template);
    return loaded;
  }

  @Test
  public void testEnvironmentSet() {
    assertEquals(VALUE1, System.getenv(KEY1), "Surefire plugin must set the variable.");
    assertEquals("", System.getenv(EMPTY), "Surefire plugin must set the variable.");
  }

  @Test
  @DisplayName("Parsing ENV variables must be explicitly enabled")
  public void testNoEnvConstructor() {
    Load loader = new Load(LoadSettings.builder().build());
    String loaded = (String) loader.loadFromString("${EnvironmentKey1}");
    assertEquals("${EnvironmentKey1}", loaded);
  }

  @Test
  @DisplayName("Parsing ENV variable which is defined and not empty")
  public void testEnvConstructor() {
    assertEquals(VALUE1, load("${EnvironmentKey1}"));
    assertEquals(VALUE1, load("${EnvironmentKey1-any}"));
    assertEquals(VALUE1, load("${EnvironmentKey1:-any}"));
    assertEquals(VALUE1, load("${EnvironmentKey1:?any}"));
    assertEquals(VALUE1, load("${EnvironmentKey1?any}"));
  }

  @Test
  @DisplayName("Parsing ENV variable which is defined as empty")
  public void testEnvConstructorForEmpty() {
    assertEquals("", load("${EnvironmentEmpty}"));
    assertEquals("", load("${EnvironmentEmpty?}"));
    assertEquals("detected", load("${EnvironmentEmpty:-detected}"));
    assertEquals("", load("${EnvironmentEmpty-detected}"));
    assertEquals("", load("${EnvironmentEmpty?detectedError}"));
    try {
      load("${EnvironmentEmpty:?detectedError}");
    } catch (MissingEnvironmentVariableException e) {
      assertEquals("Empty mandatory variable EnvironmentEmpty: detectedError", e.getMessage());
    }
  }

  @Test
  @DisplayName("Parsing ENV variable which is not set")
  public void testEnvConstructorForUnset() {
    assertEquals("", load("${EnvironmentUnset}"));
    assertEquals("", load("${EnvironmentUnset:- }"));
    assertEquals("detected", load("${EnvironmentUnset:-detected}"));
    assertEquals("detected", load("${EnvironmentUnset-detected}"));
    try {
      load("${EnvironmentUnset:?detectedError}");
    } catch (MissingEnvironmentVariableException e) {
      assertEquals("Missing mandatory variable EnvironmentUnset: detectedError", e.getMessage());
    }
    try {
      load("${EnvironmentUnset?detectedError}");
    } catch (MissingEnvironmentVariableException e) {
      assertEquals("Missing mandatory variable EnvironmentUnset: detectedError", e.getMessage());
    }
  }
}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/usecases/env/EnvVariableTest.kt
package it.krzeminski.snakeyaml.engine.kmp.usecases.env

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldEndWith
import it.krzeminski.snakeyaml.engine.kmp.api.Load
import it.krzeminski.snakeyaml.engine.kmp.api.LoadSettings
import it.krzeminski.snakeyaml.engine.kmp.env.EnvConfig
import it.krzeminski.snakeyaml.engine.kmp.exceptions.MissingEnvironmentVariableException
import it.krzeminski.snakeyaml.engine.kmp.internal.areEnvVarsSupported
import it.krzeminski.snakeyaml.engine.kmp.stringFromResources

class EnvVariableTest : FunSpec({
    // the variable is set by Gradle
    val VALUE1 = "EnvironmentValue1"
    val NULL_ENV_CONFIG = object : EnvConfig {
        override fun getValueFor(name: String, separator: String?, value: String?, environment: String?): String? {
            return null
        }
    }

    test("Parse docker-compose.yaml example").config(enabled = areEnvVarsSupported()) {
        val loader = Load(LoadSettings.builder().setEnvConfig(NULL_ENV_CONFIG).build())
        val resource = stringFromResources("/env/docker-compose.yaml")
        @Suppress("UNCHECKED_CAST")
        val compose = loader.loadOne(resource) as Map<String, Any>
        val output = compose.toString()
        output shouldEndWith "environment={URL1=EnvironmentValue1, URL2=, URL3=server3, URL4=, URL5=server5, URL6=server6}}}}"
    }

    test("Parsing ENV variables must be explicitly enabled") {
        val loader = Load()
        val loaded = loader.loadOne("\${EnvironmentKey1}") as String
        loaded shouldBe "\${EnvironmentKey1}"
    }

    fun load(template: String): String {
        val loader = Load(LoadSettings.builder().setEnvConfig(NULL_ENV_CONFIG).build())
        return loader.loadOne(template) as String
    }

    test("Parsing ENV variable which is defined and not empty").config(enabled = areEnvVarsSupported()) {
        load("\${EnvironmentKey1}") shouldBe VALUE1
        load("\${EnvironmentKey1-any}") shouldBe VALUE1
        load("\${EnvironmentKey1:-any}") shouldBe VALUE1
        load("\${EnvironmentKey1:?any}") shouldBe VALUE1
        load("\${EnvironmentKey1?any}") shouldBe VALUE1
    }

    test("Parsing ENV variable which is defined as empty").config(enabled = areEnvVarsSupported()) {
        load("\${EnvironmentEmpty}") shouldBe ""
        load("\${EnvironmentEmpty?}") shouldBe ""
        load("\${EnvironmentEmpty:-detected}") shouldBe "detected"
        load("\${EnvironmentEmpty-detected}") shouldBe ""
        load("\${EnvironmentEmpty?detectedError}") shouldBe ""
        shouldThrow<MissingEnvironmentVariableException> {
            load("\${EnvironmentEmpty:?detectedError}")
        }.also {
            it.message shouldBe "Empty mandatory variable EnvironmentEmpty: detectedError"
        }
    }

    test("Parsing ENV variable which is not set").config(enabled = areEnvVarsSupported()) {
        load("\${EnvironmentUnset}") shouldBe ""
        load("\${EnvironmentUnset:- }") shouldBe ""
        load("\${EnvironmentUnset:-detected}") shouldBe "detected"
        load("\${EnvironmentUnset-detected}") shouldBe "detected"
        shouldThrow<MissingEnvironmentVariableException> {
            load("\${EnvironmentUnset:?detectedError}")
        }.also {
            it.message shouldBe "Missing mandatory variable EnvironmentUnset: detectedError"
        }
        shouldThrow<MissingEnvironmentVariableException> {
            load("\${EnvironmentUnset?detectedError}")
        }.also {
            it.message shouldBe "Missing mandatory variable EnvironmentUnset: detectedError"
        }
    }
})
